src/fides/api/util/tcf_util.py

```diff
@@ -0,0 +1,623 @@
+    purposes/features are added to it and we modify that mapping multiple times.
+    """
+    matching_purpose_or_feature_map: Dict[int, TCFPurposeOrFeature] = {}
+
+    is_purpose_section: bool = "purposes" in tcf_component_name
+
+    for privacy_declaration_row in matching_privacy_declaration_query:
+        # Filter relevant uses or features, depending on the section of the tcf overlay
+        relevant_use_or_features: Set[str] = get_matching_data_uses_or_features(
+            record=privacy_declaration_row,
+            relevant_uses_or_features=relevant_uses_or_features,
+            is_purpose_type=is_purpose_section,
+        )
+        legal_basis_for_processing: Optional[str] = privacy_declaration_row[
+            "legal_basis_for_processing"
+        ]
+
+        for attribute in relevant_use_or_features:
+            # Add top-level entry to purpose or feature section if applicable
+            top_level_tcf_record: Optional[
+                TCFPurposeOrFeature
+            ] = _add_top_level_record_to_purpose_or_feature_section(
+                matching_purpose_or_feature_map=matching_purpose_or_feature_map,
+                is_purpose_type=is_purpose_section,
+                use_or_feature=attribute,
+            )
+
+            if not top_level_tcf_record:
+                continue
+
+            vendor_id, system_identifier = get_system_identifiers(
+                privacy_declaration_row
+            )
+            # Add top-level entry to the system section if applicable
+            if system_identifier not in system_map:
+                system_map[system_identifier] = TCFVendorRecord(
+                    id=system_identifier,  # Identify system by vendor id if it exists, otherwise use system id.
+                    name=privacy_declaration_row.system_name,
+                    description=privacy_declaration_row.system_description,
+                    has_vendor_id=bool(
+                        vendor_id
+                    ),  # Has_vendor_id will let us separate data into two sections: "tcf_vendors" and "tcf_systems"
+                )
+
+            # Copy top level purpose or feature record to preserve its initial state to use for embedded record
+            embedded_tcf_record: TCFPurposeOrFeature = top_level_tcf_record.copy()
+
+            # Override legal_bases on the embedded purpose or special purpose record
+            _extend_legal_bases(
+                embedded_tcf_record, legal_basis_for_processing, replace=True
+            )
+
+            # Embed the purpose/feature under the system if it doesn't exist, and/or consolidate legal bases
+            _embed_purpose_or_feature_under_system(
+                embedded_tcf_record=embedded_tcf_record,
+                system_section=getattr(
+                    system_map[system_identifier], tcf_component_name
+                ),
+                legal_basis_for_processing=legal_basis_for_processing,
+            )
```

Comment:
per our previous discussion, could this look something like 
```suggestion

            # Embed the purpose/feature under the system if it doesn't exist, and/or consolidate legal bases
            _embed_purpose_or_feature_under_system(
                embedded_tcf_record=top_level_tcf_record.copy(),
                system_section=getattr(
                    system_map[system_identifier], tcf_component_name
                ),
                legal_basis_for_processing=legal_basis_for_processing,
            )
```

and then see corresponding suggestion on ` _embed_purpose_or_feature_under_system` to see how i'd imagine that being tweaked...

Reply:
ah ok thanks